src/hm/HookStateVirtualNetwork.cc

```diff
@@ -0,0 +1,125 @@
+
+#include "HookStateVirtualNetwork.h"
+#include "NebulaLog.h"
+#include "SSLUtil.h"
+
+using std::string;
+
+/* -------------------------------------------------------------------------- */
+/* -------------------------------------------------------------------------- */
+
+bool HookStateVirtualNetwork::trigger(VirtualNetwork * vn)
+{
+    return vn->has_changed_state();
+}
+
+/* -------------------------------------------------------------------------- */
+/* -------------------------------------------------------------------------- */
+
+string HookStateVirtualNetwork::format_message(VirtualNetwork * vn)
+{
+    std::ostringstream oss;
+    string vn_xml;
+
+    oss << "<HOOK_MESSAGE>"
+        << "<HOOK_TYPE>STATE</HOOK_TYPE>"
+        << "<HOOK_OBJECT>NET</HOOK_OBJECT>"
+        << "<STATE>" << VirtualNetwork::state_to_str(vn->get_state()) << "</STATE>"
+        << "<RESOURCE_ID>" << vn->get_oid() << "</RESOURCE_ID>"
+        << vn->to_xml(vn_xml)
+        << "</HOOK_MESSAGE>";
+
+    string base64;
+    ssl_util::base64_encode(oss.str(), base64);
+
+    return base64;
+}
+
+/* -------------------------------------------------------------------------- */
+/* -------------------------------------------------------------------------- */
+
+int HookStateVirtualNetwork::parse_template(Template * tmpl, string& error_str)
+{
+    string state_str;
+
+    tmpl->get("STATE", state_str);
+    tmpl->erase("STATE");
+
+    state = VirtualNetwork::str_to_state(state_str);
+    if (state == VirtualNetwork::INIT)
+    {
+        error_str = "Invalid STATE: " + state_str;
+        return -1;
+    }
+
+    tmpl->add("STATE", VirtualNetwork::state_to_str(state));
+
+    return 0;
+}
+
+/* -------------------------------------------------------------------------- */
+/* -------------------------------------------------------------------------- */
+
+int HookStateVirtualNetwork::from_template(const Template * tmpl, string& error_str)
+{
+    string state_str;
+
+    if ( tmpl->get("STATE", state_str) )
+    {
+        VirtualNetwork::VirtualNetworkState _state = VirtualNetwork::str_to_state(state_str);
+
+        if (_state == VirtualNetwork::INIT)
+        {
+            error_str = "Invalid or unknown STATE attribute";
+            return -1;
+        }
+    }
+    else
+    {
+        error_str = "STATE attribute not found or invalid";
+        return -1;
+    }
+
+    return 0;
+}
+
+/* -------------------------------------------------------------------------- */
+/* -------------------------------------------------------------------------- */
+
+int HookStateVirtualNetwork::post_update_template(Template * tmpl, string& error)
+{
+    string new_state_str;
+
+    if ( tmpl->get("STATE", new_state_str) )
+    {
+        VirtualNetwork::VirtualNetworkState new_state = VirtualNetwork::str_to_state(new_state_str);
+
+        if ( new_state != VirtualNetwork::INIT )
+        {
+            state = new_state;
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

We observed that the presence of non-empty statements like `a + b` without clear context or purpose within the codebase can lead to potential confusion or misinterpretation. To mitigate this issue, it&#39;s advisable to ensure that such statements are appropriately integrated into larger constructs that contribute meaningfully to the program&#39;s logic or functionality. This could involve assigning the result of the expression to a variable.